include/veriblock/blockchain/pop/pop_state_machine.hpp

```diff
@@ -154,23 +151,17 @@ struct PopStateMachine {
         }
 
       }  // end for
-
-      // since we have successfully applied the block, clear BLOCK_FAILED_POP
-      ed_.revalidateSubtree(index, BLOCK_FAILED_POP, /*do fr=*/false);
-
-    } else {
-      VBK_ASSERT(!index.hasFlags(BLOCK_FAILED_POP) &&
-                 "state corruption: an empty block must not be invalid due to "
-                 "its payloads");
     }
 
     // if the applied block count equals the size of the chain between the root
     // and index, we have just applied a block on top of the only applied chain,
     // so it is fully valid
-    if (index.pprev->hasFlags(BLOCK_CAN_BE_APPLIED) &&
+    if (index.pprev->isValid(BLOCK_CAN_BE_APPLIED) &&
         index.getHeight() ==
             ed_.getRoot().getHeight() + ed_.appliedBlockCount) {
-      index.setFlag(BLOCK_CAN_BE_APPLIED);
+      index.raiseValidity(BLOCK_CAN_BE_APPLIED);
+    } else {
+      index.raiseValidity(BLOCK_HAS_BEEN_APPLIED);
     }
```

Comment:
This else statement looks weird. So if ed_.appliedBlockCount is less than index.height block will get BLOCK_HAS_BEEN_APPLIED status?

Reply:
not sure I understand the question. this functionality isn't changed by this PR

Reply:
Ok I see that BLOCK_CAN_BE_APPLIED is higher validity than BLOCK_HAS_BEEN_APPLIED. So the logic here is to set BLOCK_HAS_BEEN_APPLIED first and set BLOCK_CAN_BE_APPLIED if the previous chain is fully applied.
The logic seems OK but was hard to embrace cause is looks like BLOCK_CAN_BE_APPLIED should come before BLOCK_HAS_BEEN_APPLIED.